blockstore/blockstore.go

```diff
@@ -216,7 +230,15 @@ func (bs *blockstore) Put(ctx context.Context, block blocks.Block) error {
 			return nil // already stored.
 		}
 	}
-	return bs.datastore.Put(ctx, k, block.RawData())
+	if err := bs.datastore.Put(ctx, k, block.RawData()); err != nil {
+		return err
+	}
+
+	if bs.provider != nil {
+		logger.Debugf("blockstore: provide %s", block.Cid())
+		return bs.provider.Provide(ctx, block.Cid(), true)
```

Comment:
Maybe we should ignore the error returned by `bs.provider.Provide()`. I am not too familiar with the `blockstore`, but by seeing a few use cases of `Put()`, we probably don't want to consider a `blockstore.Put()` as a failure if the cid couldn't be advertised.

For instance, [`blockservice.getBlock`](https://github.com/ipfs/boxo/blob/803408a01b73ae23f2a62521439e5828d93a3a7f/blockservice/blockservice.go#L270-L274) calls `blockstore.Put()`, and propagates any potential error. I think that `blockservice.getBlock()` should succeed, even though the provide could fail.

Also, `Put()` doesn't need to block on the `Provide()` operation.

```suggestion
		go bs.provider.Provide(ctx, block.Cid(), true)
```

Reply:
Agreed on errors, disagree on not-blocking for Provide.

If you're provider is the reprovider, it just queues and returns. If you're provider is the (current) DHT, this call will effectively block for several seconds. I think it is ok that it does, because the alternative is launching goroutines (1000 blocks added -> 1000 concurrent provide operations goroutines). So for me the responsibility of handling concurrent provides is in the Provider implementation and the blockstore should not take the liberty to handle concurrency there.

Reply:
Ok, I agree with your reasoning 👍🏻